[TRTLLM-6291] feat: Add user-provided speculative decoding support

Funatiq opened the pull request:
## Description

- Introduced UserProvidedConfig for user-defined speculative decoding configurations.
- Updated SpeculativeDecodingMode to include USER_PROVIDED option.
- Enhanced get_spec_resource_manager and get_spec_drafter functions to handle user-provided drafter.
- Modified PyExecutor and related classes to accommodate the new user-provided decoding feature.
- Added unit tests to validate user-provided decoding functionality.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

AlessioNetti commented on tensorrt_llm/_torch/pyexecutor/py_executor_creator.py:
```diff
@@ -322,8 +323,13 @@ def create_py_executor(executor_config: ExecutorConfig,
-        spec_resource_manager = get_spec_resource_manager(
-            spec_config, model_engine.model.config, model_engine.batch_size * 2)
+        if isinstance(spec_config, UserProvidedDecodingConfig):
+            spec_resource_manager = pytorch_backend_config.extra_resource_managers.pop(
```
It might be a good idea here to check if `"spec_resource_manager"` actually is in the `extra_resource_managers` dict to avoid a `KeyError`. One should also expose the  `"spec_resource_manager"` key as an enum/constant.

AlessioNetti commented on tensorrt_llm/_torch/speculative/ngram.py:
```diff
@@ -108,6 +108,10 @@ def prepare_resources(self, scheduled_batch: ScheduledRequests):
                 pad_length = self.max_num_draft_tokens - len(draft_tokens)
                 draft_tokens.extend([request.py_end_id] * pad_length)
+            else:
+                draft_tokens = [request.py_end_id] * self.max_num_draft_tokens
```
I suppose this is part of the code that can/should be overridden, but doesn't this assignment mean that the model will still try to verify a dummy draft even if none can be found, leading to (potential) performance loss?

AlessioNetti commented on tests/unittest/_torch/speculative/test_user_provided.py:
```diff
@@ -0,0 +1,93 @@
+
+# TODO: Add cuda graph enabled tests.
+# Cuda graph cannot currently be enabled for ngram because cuda graph requires
+# spec metadata and ngram does not have it.
```
Getting CUDA graphs support for the ngram drafter would be quite important, I think, as not having that offsets most of the latency gains gained with speculative decoding.

Funatiq committed:
refactor: Use keyword arguments for clarity

- Updated the create_py_executor function to pass parameters as keyword arguments for better readability.
- Modified the PyTorchModelEngine instantiation to use named parameters for model_path and pytorch_backend_config.
- Added a keyword-only argument to create_py_executor_instance for consistency in parameter handling.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Remove unused SampleState parameter from drafter methods

- Updated the prepare_draft_tokens method in Drafter and NGramDrafter classes to remove the unused SampleState parameter.
- Adjusted the PyExecutor class to call prepare_draft_tokens without the SampleState argument.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
feat: Add user-provided speculative decoding support

- Introduced UserProvidedConfig for user-defined speculative decoding configurations.
- Updated SpeculativeDecodingMode to include USER_PROVIDED option.
- Enhanced get_spec_resource_manager and get_spec_drafter functions to handle user-provided drafter.
- Modified PyExecutor and related classes to accommodate the new user-provided decoding feature.
- Added unit tests to validate user-provided decoding functionality.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10500](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10500/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #7774](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7774/) completed with status: 'FAILURE'

Funatiq committed:
refactor: Simplify NGramSpecMetadata

- Removed the NGramSpecMetadata class, consolidating its functionality into SpecMetadata.
- Updated imports in ngram.py and utils.py to reflect the removal of NGramSpecMetadata.
- Added a placeholder __post_init__ method in SpecMetadata as a default implementation.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
fix: Update and unwaive ngram test case

- Removed a test_ngram test case from waives.txt.
- Updated the batch sizes parameter in the CudaGraphConfig in test_ngram.py.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
test: Change model to Meta-Llama-3.1-8B

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "B200_PCIe-PyTorch-1, B200_PCIe-PyTorch-2, B200_PCIe-PyTorch-3"

tensorrt-cicd commented:
[PR_Github #10588](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10588/) [ run ] triggered by Bot

Funatiq commented:
/bot run

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10791](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10791/) [ run ] triggered by Bot

wili-65535 reviewed (approved):
+ LGTM!
+ Good improvement on the Drafter construction ("construct in resource manager and pass to get_spec_drafter()" to "construct in get_spec_drafter() and pass to resource manager"), making the logic clearer.
+ Remove input argument `state` from `Drafterr.prepare_draft_tokens()`. this is remained for overlap-scheduler, we do not need it at this moment.
+ Maybe we need document update (in a later PR). For example, user need to implement their own `prepare_draft_tokens()` in `class UserProvidedConfig`.

tensorrt-cicd commented:
[PR_Github #11012](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11012/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11019](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11019/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8141](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8141/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/8141/test-results/rerun_report.html) for details.

mikeiovine reviewed (approved):
Only have nitpicks, thanks

Funatiq commented on tensorrt_llm/_torch/pyexecutor/model_engine.py:
```diff
@@ -341,6 +341,7 @@ class PyTorchModelEngine(ModelEngine):
 
     def __init__(
         self,
+        *,
```
Because there are actually a lot here. I could have added it below `pytorch_backend_config` but why not go all the way? The IDE should avoid any annoyances.
